File: tema5_cross_platform/server_csharp/musicFestival/protobuf/ProtoUtils.cs
namespace Proto
{
    public class ProtoUtils {

    public static Request createLoginRequest(User user)
    {
        Proto.User userDTO = new Proto.User {Username = user.Username, Password = user.Password};
        Request request = new Request {Type = Request.Types.Type.Login, User = userDTO};
        return request;
    }

    public static Request createLogoutRequest(csharpMusicFestival.domain.User user)
    {
        Proto.User userDTO = new Proto.User {Username = user.Name, Password = user.Password};
        Request request = new Request{Type = Request.Types.Type.Logout, User = userDTO};
        return request;
    }

    public static Response createGetShowsResponse(Show[] shows){
        Response response = new Response { 
            Type = Response.Types.Type.GetShows
        };
        foreach (var show in shows)
        {
            Proto.Show showDTO = new Proto.Show
            {
                Id = show.Id,
                ArtistName = show.ArtistName,
                AvailableTicketsNumber = show.AvailableTicketsNumber,
                Date = show.Date,
                Location = show.Location,
                SoldTicketsNumber = show.SoldTicketsNumber
            };
            response.Shows.Add(showDTO);
        }
        return response;
    }

    public static Response createGetArtistsByDateResponse(Artist[] artists){
        Response response = new Response { 
            Type = Response.Types.Type.GetArtistsByDate
        };
        foreach (var artist in artists)
        {
            Proto.Artist artistDTO = new Proto.Artist
            {
                Name = artist.Name,
                AvailableTicketsNumber = artist.AvailableTicketsNumber,
                Date = artist.Date,
                Location = artist.Location,
            };
            response.Artist.Add(artistDTO);
        }
        return response;
    }

    public static Response createOkResponse()
    {
        Response response = new Response{ Type=Response.Types.Type.Ok};
        return response;
    }

    public static Response createErrorResponse(string text)
    {
        Response response = new Response{
            Type=Response.Types.Type.Error, Error=text};
        return response;
    }

    public static string GetError(Response response){
        string errorMessage = response.Error;
        return errorMessage;
    }

    public static csharpMusicFestival.domain.User GetUser(Response response){
        csharpMusicFestival.domain.User user = new csharpMusicFestival.domain.User(response.User.Username, response.User.Password);
        return user;
    }

    public static csharpMusicFestival.domain.User GetUser(Request request)
    {
        var user = new csharpMusicFestival.domain.User {Name = request.User.Username, Password = request.User.Password};
        return user;
    }

    public static string GetDate(Request request)
    {
        return request.Date;
    }

    public static Ticket GetTicket(Response response)
    {
        Ticket ticket = new Ticket
        {
            ShowId = response.Ticket.ShowId,
            Number = response.Ticket.Number,
            PurchaserName = response.Ticket.PurchaserName
        };
        return ticket;
    }

    public static csharpMusicFestival.domain.Ticket GetTicket(Request request)
    {
        var ticket = new csharpMusicFestival.domain.Ticket
        {
            ShowId = request.Ticket.ShowId,
            Number = request.Ticket.Number,
            PurchaserName = request.Ticket.PurchaserName
        };
        return ticket;
    }

    public static Show[] GetShows(Response response) {
        Show[] shows = new Show[response.Shows.Count];
        for(int i = 0; i < response.Shows.Count;i++)
        {
            Show show = new Show
            {
                Id = response.Shows[i].Id,
                Location = response.Shows[i].Location,
                Date = response.Shows[i].Date,
                ArtistName = response.Shows[i].ArtistName,
                AvailableTicketsNumber = response.Shows[i].AvailableTicketsNumber,
                SoldTicketsNumber = response.Shows[i].SoldTicketsNumber
            };
            shows[i] = show;
        }
        return shows;
    }

    public static Artist[] GetArtists(Response response) {
        Artist[] artists = new Artist[response.Artist.Count];
        for(int i = 0; i < response.Artist.Count;i++)
        {
            Artist artist = new Artist
            {
                Location = response.Shows[i].Location,
                Date = response.Shows[i].Date,
                Name = response.Shows[i].ArtistName,
                AvailableTicketsNumber = response.Shows[i].AvailableTicketsNumber,
            };
            artists[i] = artist;
        }
        return artists;
    }

    public static Response CreateGetShowsResponse(csharpMusicFestival.domain.Show[] shows)
    {
        Response response = new Response { 
            Type=Response.Types.Type.GetShows
        };
        foreach (csharpMusicFestival.domain.Show show in shows)
        {
            Proto.Show showDTO = new Proto.Show
            {
                Id = show.Id,
                Location = show.Location,
                Date = show.Date,
                ArtistName = show.ArtistName,
                AvailableTicketsNumber = show.AvailableTicketsNumber,
                SoldTicketsNumber = show.SoldTicketsNumber
            };
            response.Shows.Add(showDTO);
        }

        return response;
    }

    public static Response CreateGetArtistsByDateResponse(csharpMusicFestival.domain.Artist[] artists)
    {
        Response response = new Response { 
            Type=Response.Types.Type.GetArtistsByDate
        };
        foreach (csharpMusicFestival.domain.Artist artist in artists)
        {
            Proto.Artist artistDTO = new Proto.Artist
            {
                Location = artist.Location,
                Date = artist.Date,
                Name = artist.Name,
                AvailableTicketsNumber = artist.AvailableTicketsNumber,
            };
            response.Artist.Add(artistDTO);
        }

        return response;
    }

    public static Response CreateBuyTicketResponse(csharpMusicFestival.domain.Ticket ticket)
    {
        var ticketDTO = new Proto.Ticket
        {
            ShowId = ticket.ShowId,
            Number = ticket.Number,
            PurchaserName = ticket.PurchaserName
        };

        Response response = new Response { Type = Response.Types.Type.BuyTicket, Ticket = ticketDTO};
        return response;
    }
    }
}

File: tema5_cross_platform/client_java/Networking/src/main/java/ProtoUtils.java
import domain.Artist;
import domain.Show;
import domain.Ticket;
import domain.User;
import protobuffprotocol.Protobufs;

public class ProtoUtils {

    public static Protobufs.Request createLoginRequest(User user){
        Protobufs.User userDTO = Protobufs.User.newBuilder().setUsername(user.getUsername()).setPassword(user.getPassword()).build();
        return Protobufs.Request.newBuilder().setType(Protobufs.Request.Type.Login)
                .setUser(userDTO).build();
    }

    public static Protobufs.Request createLogoutRequest(User user){
        Protobufs.User userDTO = Protobufs.User.newBuilder().setUsername(user.getUsername()).setPassword(user.getPassword()).build();
        return Protobufs.Request.newBuilder().setType(Protobufs.Request.Type.Logout)
                .setUser(userDTO).build();
    }

    public static Protobufs.Response createGetShowsResponse(Show[] shows){
        Protobufs.Response.Builder response = Protobufs.Response.newBuilder()
                .setType(Protobufs.Response.Type.GetShows);
        for (Show show : shows){
            Protobufs.Show showDTO = Protobufs.Show.newBuilder()
                    .setId(show.getId())
                    .setArtistName(show.getArtistName())
                    .setAvailableTicketsNumber(show.getAvailableTicketsNumber())
                    .setDate(show.getDate())
                    .setLocation(show.getLocation())
                    .setSoldTicketsNumber(show.getSoldTicketsNumber())
                    .build();
            response.addShows(showDTO);
        }
        return response.build();
    }

    public static Protobufs.Response createGetArtistsByDateResponse(Artist[] artists){
        Protobufs.Response.Builder response = Protobufs.Response.newBuilder()
                .setType(Protobufs.Response.Type.GetArtistsByDate);
        for (Artist artist : artists){
            Protobufs.Artist artistDTO = Protobufs.Artist.newBuilder()
                    .setName(artist.getName())
                    .setAvailableTicketsNumber(artist.getAvailableTicketsNumber())
                    .setDate(artist.getDate())
                    .setLocation(artist.getLocation())
                    .build();
            response.addArtist(artistDTO);
        }
        return response.build();
    }

    public static Protobufs.Response createOkResponse(){
        return Protobufs.Response.newBuilder()
                .setType(Protobufs.Response.Type.Ok).build();
    }

    public static Protobufs.Response createErrorResponse(String text){
        return Protobufs.Response.newBuilder()
                .setType(Protobufs.Response.Type.Error)
                .setError(text).build();
    }

    public static String getError(Protobufs.Response response){
        return response.getError();
    }

    public static User getUser(Protobufs.Response response){
        User user=new User();
        user.setUsername(response.getUser().getUsername());
        return user;
    }

    public static User getUser(Protobufs.Request request){
        User user=new User();
        user.setUsername(request.getUser().getUsername());
        user.setPassword(request.getUser().getPassword());
        return user;
    }

    public static String getDate(Protobufs.Request request) {
        return request.getDate();
    }

    public static Ticket getTicket(Protobufs.Response response) {
        Ticket ticket = new Ticket();
        ticket.setShowId(response.getTicket().getShowId());
        ticket.setNumber(response.getTicket().getNumber());
        ticket.setPurchaserName(response.getTicket().getPurchaserName());
        return ticket;
    }

    public static Ticket getTicket(Protobufs.Request request) {
        Ticket ticket = new Ticket();
        ticket.setShowId(request.getTicket().getShowId());
        ticket.setNumber(request.getTicket().getNumber());
        ticket.setPurchaserName(request.getTicket().getPurchaserName());
        return ticket;
    }

    public static Show[] getShows(Protobufs.Response response) {
        Show[] shows = new Show[response.getShowsCount()];
        for(int i = 0; i < response.getShowsCount();i++){
            Protobufs.Show showDTO = response.getShows(i);
            Show show = new Show();
            show.setId(showDTO.getId());
            show.setLocation(showDTO.getLocation());
            show.setDate(showDTO.getDate());
            show.setArtistName(showDTO.getArtistName());
            show.setAvailableTicketsNumber(showDTO.getAvailableTicketsNumber());
            show.setSoldTicketsNumber(showDTO.getSoldTicketsNumber());
            shows[i] = show;
        }
        return shows;
    }

    public static Artist[] getArtists(Protobufs.Response response) {
        Artist[] artists = new Artist[response.getArtistCount()];
        for(int i = 0; i < response.getArtistCount();i++){
            Protobufs.Artist artistDTO = response.getArtist(i);
            Artist artist = new Artist();
            artist.setName(artistDTO.getName());
            artist.setAvailableTicketsNumber(artistDTO.getAvailableTicketsNumber());
            artist.setDate(artistDTO.getDate());
            artist.setLocation(artistDTO.getLocation());
            artists[i] = artist;
        }
        return artists;
    }

    public static Protobufs.Request getShowsRequest() {
        return Protobufs.Request.newBuilder().setType(Protobufs.Request.Type.GetShows).build();
    }

    public static Protobufs.Request getArtistsByDateRequest(String date) {
        return Protobufs.Request.newBuilder().setType(Protobufs.Request.Type.GetArtistsByDate).setDate(date).build();
    }

    public static Protobufs.Request buyTicketRequest(Ticket ticket) {
        Protobufs.Ticket ticketDTO = Protobufs.Ticket.newBuilder()
                .setShowId(ticket.getShowId())
                .setPurchaserName(ticket.getPurchaserName())
                .setNumber(ticket.getNumber())
                .build();
        return Protobufs.Request.newBuilder().setType(Protobufs.Request.Type.BuyTicket).setTicket(ticketDTO).build();
    }

    public static Protobufs.Response createTicketSoldResponse(Ticket ticket) {
        Protobufs.Ticket ticketDTO = Protobufs.Ticket.newBuilder()
                .setShowId(ticket.getShowId())
                .setPurchaserName(ticket.getPurchaserName())
                .setNumber(ticket.getNumber())
                .build();

        return Protobufs.Response.newBuilder()
                .setType(Protobufs.Response.Type.BuyTicket)
                .setTicket(ticketDTO).build();
    }
}
